# GENOMIC PREDICTION FOR LONGITUDINAL RESIDUAL FEED INTAKE -

## **COMPUTING CORRECTED PHENOTYPES**






This program is used to compute the corrected phenotypes obtained with different models Random regression, SAD and Multiple-trait models

In [1]:
# Load required packages
library(pacman)
pacman::p_load(dplyr,reshape2, stringr, plyr, psych, corrplot)



# set path work drive
#path=getwd()
setwd("C:/Users/HP/OneDrive/RFI_all/RFI_all")


# Read data 
pheno <- read.table("pheno1.dat", quote="\"", comment.char="")


# renames columns
names(pheno)<-c("animal","identif" ,"bande",
                "sex" ,"lignee"  ,"gener",
                "loge"  , "eleve","pds_dc",  
                "pds_fc" ,  "age" ,  "BW","BWpre",
                "age_dc",
                "temps","FCR","ADG","AMBW"  ,
                "BF","FI" ,"RFI","week","pe1",
                "pe2","pe3",
                "selec1","selec2","e_RR",
                "e_RRss","e_SAD","e_SADss","e_MTA","e_MTH")

pheno <- select(pheno, -c(pe1,pe3)) # drop unnecessary columns


This dataset contains the following information:
- animal, identif = id of animal
- BW = Body weight
- BWpre = predicted BW using a linear interpolation
- temps = time (in week, from week 4 to week 13 from the beginning of the test)
- pe1 = period 1 (week 1-2)
- pe2 = period 2 (week 3-5)
- e_  = residual effects


 **Display the first lines of dataset 'pheno'**

In [2]:
head(pheno)

animal        identif       bande  sex lignee gener loge eleve pds_dc pds_fc
1 17MAG20005072 17MAG20005072 200011 2   n      G0    72   17MAG 24     128   
2 17MAG20005072 17MAG20005072 200011 2   n      G0    72   17MAG 24     128   
3 17MAG20005072 17MAG20005072 200011 2   n      G0    72   17MAG 24     128   
4 17MAG20005072 17MAG20005072 200011 2   n      G0    72   17MAG 24     128   
5 17MAG20005072 17MAG20005072 200011 2   n      G0    72   17MAG 24     128   
6 17MAG20005072 17MAG20005072 200011 2   n      G0    72   17MAG 24     128   
  ... week pe2 selec1 selec2 e_RR    e_RRss  e_SAD      e_SADss    e_MTA 
1 ... 1    1   0      1      -31.580 -30.980 -0.0009618 -9.739e-04 -28.78
2 ... 2    1   0      1       -4.791  -4.344 -0.0001196 -9.429e-05  -1.39
3 ... 3    2   0      1       16.910  17.230  0.0002545  2.841e-04  19.38
4 ... 4    2   0      1       16.470  16.690  0.0006243  6.332e-04  18.52
5 ... 5    2   0      1      -60.050 -59.900 -0.0016420 -1.647e-03 -58.44
6 ... 6    3   0      1       13.970  14.090  0.0004870  4.940e-04  11.81
  e_MTH  
1 -28.560
2  -1.137
3  20.080
4  19.210
5 -57.750
6  11.600

In [3]:
length(unique(pheno$animal))

[1] 2397

In above file contains data of 2397 animals, including batch(bande), sex, line (lignee), loge, eleve, BW at the begining and BW_at the end of test

 # Random regression models with matrix A

In [4]:
# Read data (outputs of Asreml program)
sln_RRAe <- read.table("ide_RRA",quote="\"") # Permanental random effect
sln_RRAg <- read.table("sln_RRA",quote="\"") # Genetic random effect
sln_RRAg$V3=sln_RRAe$V3 + sln_RRAg$V3

In [5]:
sln_RRAg = cbind(sln_RRAg,as.data.frame(matrix(unlist( strsplit(as.character(sln_RRAg$V2)  , "\\.")) , ncol = 2 , byrow = TRUE )))
names(sln_RRAg)<-c("kk","a","para","SE","gr","animal")
head(sln_RRAg)

kk                  a               para       SE    gr animal       
1 leg(temps,2).animal 1.17M3019830268 -0.0459300 8.166 1  17M3019830268
2 leg(temps,2).animal 1.17M3019830280  0.0009301 8.166 1  17M3019830280
3 leg(temps,2).animal 1.17M3019830492 -0.0234300 8.166 1  17M3019830492
4 leg(temps,2).animal 1.17M3019830508 -0.0234300 8.166 1  17M3019830508
5 leg(temps,2).animal 1.17M3019840010 -0.0215700 8.166 1  17M3019840010
6 leg(temps,2).animal 1.17M3019840036 -0.0468600 8.166 1  17M3019840036

In [6]:
sln_RRAg$para1=ifelse(sln_RRAg$gr==1,sln_RRAg$para,0)  
sln_RRAg$para2=ifelse(sln_RRAg$gr==2,sln_RRAg$para,0) 
sln_RRAg$para3=ifelse(sln_RRAg$gr==3,sln_RRAg$para,0)  
sln_RRAg=arrange(sln_RRAg,animal,gr)  
head(sln_RRAg)

kk                  a               para      SE    gr animal        para1   
1 leg(temps,2).animal 1.0000019851067  0.019560 8.166 1  0000019851067  0.01956
2 leg(temps,2).animal 2.0000019851067  0.035280 7.498 2  0000019851067  0.00000
3 leg(temps,2).animal 3.0000019851067  0.006404 4.539 3  0000019851067  0.00000
4 leg(temps,2).animal 1.0000019858909 -0.152500 8.165 1  0000019858909 -0.15250
5 leg(temps,2).animal 2.0000019858909 -0.161000 7.497 2  0000019858909  0.00000
6 leg(temps,2).animal 3.0000019858909 -0.025930 4.538 3  0000019858909  0.00000
  para2    para3    
1  0.00000  0.000000
2  0.03528  0.000000
3  0.00000  0.006404
4  0.00000  0.000000
5 -0.16100  0.000000
6  0.00000 -0.025930

In [7]:
tatemp <- read.table("OP.txt",quote="\"") 
names(tatemp)[1]<-"temps"  
names(tatemp)[2]<-"f1"  
names(tatemp)[3]<-"f2"  
names(tatemp)[4]<-"f3"  
Temp=t(as.matrix(tatemp[2:4]))


In [8]:
# matix T
head(Temp)

[,1]     [,2]     [,3]     [,4]     [,5]     [,6]     [,7]     [,8]    
f1  0.70711  0.70711  0.70711  0.70711  0.70711  0.70711  0.70711  0.70711
f2 -1.22474 -0.95258 -0.68041 -0.40825 -0.13608  0.13608  0.40825  0.68041
f3  1.58114  0.64417 -0.05856 -0.52705 -0.76129 -0.76129 -0.52705 -0.05856
   [,9]    [,10]  
f1 0.70711 0.70711
f2 0.95258 1.22474
f3 0.64417 1.58114

In [9]:
#sln1=merge(sln,pheno,by = "animal")

SLN=sln_RRAg[,c("animal","gr","para")]
SLN1=arrange(SLN,animal,gr)
#SLN2=reshape(SLN1, timevar = "gr",idvar = "animal",direction="wide",sep="")
SLN2=reshape(SLN1,varying = NULL, timevar = "gr",idvar = "animal",direction="wide",sep="")
SLN3=arrange(SLN2,animal)




SLN4=SLN3[,2:4]
SLN5=as.matrix(SLN4)
EBV_all=as.data.frame((SLN5)%*%(Temp))
EBV_all$animal=SLN2$animal


df_effect=melt(EBV_all, id.vars = "animal", 
               measure.vars = c("V1", "V2","V3",
                                "V4","V5","V6","V7","V8","V9","V10"))

df_effect = arrange(df_effect,animal,variable)
df_effect$week=c(rep(1:10,3986))

In [10]:
head(df_effect)

animal        variable value        week
1 0000019851067 V1       -0.019252135 1   
2 0000019851067 V2       -0.015650686 2   
3 0000019851067 V3       -0.010548811 3   
4 0000019851067 V4       -0.003947217 4   
5 0000019851067 V5        0.004154868 5   
6 0000019851067 V6        0.013756673 6

In [11]:
pheno1=merge(df_effect,pheno,by=c("animal","week"))
pheno2=arrange(pheno1,animal,week)
pheno2$y_RR = pheno2$e_RR +pheno2$value

pheno2=subset(pheno2,select=-c(variable,value))
write.table(pheno2,"dat_correct_RRA")

In [12]:
head(pheno2)
summary(pheno2)

animal        week identif       bande  sex lignee gener loge eleve pds_dc
1 17MAG20005070 2    17MAG20005070 200011 2   n      G0    71   17MAG 25    
2 17MAG20005070 4    17MAG20005070 200011 2   n      G0    71   17MAG 25    
3 17MAG20005070 5    17MAG20005070 200011 2   n      G0    71   17MAG 25    
4 17MAG20005070 6    17MAG20005070 200011 2   n      G0    71   17MAG 25    
5 17MAG20005070 7    17MAG20005070 200011 2   n      G0    71   17MAG 25    
6 17MAG20005070 8    17MAG20005070 200011 2   n      G0    71   17MAG 25    
  ... pe2 selec1 selec2 e_RR    e_RRss  e_SAD      e_SADss    e_MTA   e_MTH  
1 ... 1   0      1       74.510  74.850  1.718e-03  1.719e-03  74.170  74.060
2 ... 2   0      1       67.060  67.430  1.574e-03  1.593e-03  63.490  64.150
3 ... 2   0      1      -18.340 -17.980 -6.276e-04 -6.220e-04 -27.610 -26.970
4 ... 3   0      1      -29.070 -28.730 -4.628e-04 -4.519e-04 -17.430 -17.220
5 ... 3   0      1       -6.171  -5.867  1.343e-05  2.321e-05  -2.308  -2.115
6 ... 3   0      1       -8.770  -8.502 -1.026e-04 -1.043e-04 -13.520 -13.330
  y_RR     
1  94.97610
2  79.52834
3 -11.37910
4 -28.62189
5 -13.24151
6 -24.36453

           animal           week                 identif          bande       
 17MAG20005072:   10   Min.   : 1.000   17MAG20005072:   10   Min.   :200011  
 17MAG20005146:   10   1st Qu.: 3.000   17MAG20005146:   10   1st Qu.:200406  
 17MAG20005197:   10   Median : 5.000   17MAG20005197:   10   Median :200612  
 17MAG20005251:   10   Mean   : 5.468   17MAG20005251:   10   Mean   :200613  
 17MAG20005297:   10   3rd Qu.: 8.000   17MAG20005297:   10   3rd Qu.:200812  
 17MAG20005301:   10   Max.   :10.000   17MAG20005301:   10   Max.   :201111  
 (Other)      :22893                    (Other)      :22893                   
      sex        lignee        gener           loge          eleve      
 Min.   :1.000   -:11744   G7     :4970   Min.   : 71.00   17MAG:13355  
 1st Qu.:2.000   +:10456   G6     :3748   1st Qu.: 81.00   86001: 9598  
 Median :2.000   n:  753   G5     :3218   Median : 91.00                
 Mean   :2.016             G4     :3107   Mean   : 87.89                
 3r

## Random regression models with matrix H

In [13]:
# Random regression model with matrix H
#rm(list = ls()) # ecrase previous memories
###############################################################################################
pheno2=read.table("dat_correct_RRA")


sln_RRHe <- read.table("ide_RRH",quote="\"") # Permanental random effect
sln_RRHg <- read.table("sln_RRH",quote="\"") # Genetic random effect
sln_RRHg$V3=sln_RRHe$V3 + sln_RRHg$V3

# sln <- read.table("C:/Users/vhtran/Google Drive/these/RFI/INRA/validation/ide_RRA",quote="\"") # 
# sln_1 <- read.table("C:/Users/vhtran/Google Drive/these/RFI/INRA/validation/sln_RRA",quote="\"") 
m_RRH <- as.data.frame(matrix( unlist( strsplit(as.character(sln_RRHg$V2)  , "\\." ) ) , ncol = 2 , byrow = TRUE ))
sln_RRHg=cbind(sln_RRHg,m_RRH)
names(sln_RRHg)<-c("kk","a","para","SE","gr","animal")



In [14]:
head(sln_RRHg)

kk                   a               para     SE    gr animal       
1 leg(temps,2).giv(ani 1.17M3019830268 -0.11520 8.057 1  17M3019830268
2 leg(temps,2).giv(ani 1.17M3019830280 -0.00814 8.057 1  17M3019830280
3 leg(temps,2).giv(ani 1.17M3019830492 -0.05354 8.057 1  17M3019830492
4 leg(temps,2).giv(ani 1.17M3019830508 -0.05354 8.057 1  17M3019830508
5 leg(temps,2).giv(ani 1.17M3019840010 -0.06982 8.057 1  17M3019840010
6 leg(temps,2).giv(ani 1.17M3019840036 -0.10710 8.057 1  17M3019840036

In [15]:
sln_RRHg$para1=ifelse(sln_RRHg$gr==1,sln_RRHg$para,0)  
sln_RRHg$para2=ifelse(sln_RRHg$gr==2,sln_RRHg$para,0) 
sln_RRHg$para3=ifelse(sln_RRHg$gr==3,sln_RRHg$para,0)  
sln_RRHg=arrange(sln_RRHg,animal,gr)  
head(sln_RRHg)


kk                   a               para      SE    gr animal       
1 leg(temps,2).giv(ani 1.0000019851067  0.018410 8.057 1  0000019851067
2 leg(temps,2).giv(ani 2.0000019851067  0.031990 7.319 2  0000019851067
3 leg(temps,2).giv(ani 3.0000019851067  0.004749 3.771 3  0000019851067
4 leg(temps,2).giv(ani 1.0000019858909 -0.243100 8.059 1  0000019858909
5 leg(temps,2).giv(ani 2.0000019858909 -0.397000 7.320 2  0000019858909
6 leg(temps,2).giv(ani 3.0000019858909  0.016640 3.771 3  0000019858909
  para1    para2    para3   
1  0.01841  0.00000 0.000000
2  0.00000  0.03199 0.000000
3  0.00000  0.00000 0.004749
4 -0.24310  0.00000 0.000000
5  0.00000 -0.39700 0.000000
6  0.00000  0.00000 0.016640

In [16]:
SLN_RRH=sln_RRHg[,c("animal","gr","para")]
SLN_RRH1=arrange(SLN_RRH,animal,gr)
#SLN2=reshape(SLN1, timevar = "gr",idvar = "animal",direction="wide",sep="")
SLN_RRH2=reshape(SLN_RRH1,varying = NULL, timevar = "gr",idvar = "animal",direction="wide",sep="")
SLN_RRH3=arrange(SLN_RRH2,animal)



SLN_RRH4=SLN_RRH3[,2:4]
SLN_RRH5=as.matrix(SLN_RRH4)
EBV_all1=as.data.frame((SLN_RRH5)%*%(Temp))
EBV_all1$animal=SLN_RRH2$animal

In [17]:
head(EBV_all1)

V1          V2          V3           V4           V5          V6         
1 -0.01865270 -0.01439598 -0.009026522 -0.002544983  0.00504933  0.01375573
2  0.34063351  0.21699481  0.097249891 -0.018593303 -0.13054255 -0.23859007
3 -0.01865270 -0.01439598 -0.009026522 -0.002544983  0.00504933  0.01375573
4  0.05959042  0.03349862  0.010659931 -0.008924683 -0.02525605 -0.03833333
5  0.26493798  0.21916376  0.175319683  0.133408547  0.09342764  0.05537967
6  0.26493798  0.21916376  0.175319683  0.133408547  0.09342764  0.05537967
  V7          V8          V9          V10         animal       
1  0.02357485  0.03450611  0.04655009  0.05970616 0000019851067
2 -0.34274380 -0.44299565 -0.53935371 -0.63181005 0000019858909
3  0.02357485  0.03450611  0.04655009  0.05970616 0000019858971
4 -0.04815751 -0.05472747 -0.05804432 -0.05810709 0000019859155
5  0.01926185 -0.01492295 -0.04717761 -0.07749932 0000019860769
6  0.01926185 -0.01492295 -0.04717761 -0.07749932 0000019860909

In [18]:
df_effect1=melt(EBV_all1, 
               id.vars = "animal", 
               measure.vars = c("V1", "V2","V3","V4","V5","V6","V7","V8","V9","V10"))

df_effect1= arrange(df_effect1,animal,variable)
df_effect1$week=c(rep(1:10,3986))

# combine the last dataset pheno2 with EBV from RR with matrix H
pheno3=merge(df_effect1,pheno2,by=c("animal","week"))
pheno4=arrange(pheno3,animal,week)
pheno4$y_RRH = pheno4$e_RRss+pheno4$value
pheno4=subset(pheno4,select=-c(variable,value))

#write.table(pheno4,"dat_correct_RRA_RRH")

In [19]:
head(pheno4)

animal        week identif       bande  sex lignee gener loge eleve pds_dc
1 17MAG20005070 2    17MAG20005070 200011 2   n      G0    71   17MAG 25    
2 17MAG20005070 4    17MAG20005070 200011 2   n      G0    71   17MAG 25    
3 17MAG20005070 5    17MAG20005070 200011 2   n      G0    71   17MAG 25    
4 17MAG20005070 6    17MAG20005070 200011 2   n      G0    71   17MAG 25    
5 17MAG20005070 7    17MAG20005070 200011 2   n      G0    71   17MAG 25    
6 17MAG20005070 8    17MAG20005070 200011 2   n      G0    71   17MAG 25    
  ... selec1 selec2 e_RR    e_RRss  e_SAD      e_SADss    e_MTA   e_MTH  
1 ... 0      1       74.510  74.850  1.718e-03  1.719e-03  74.170  74.060
2 ... 0      1       67.060  67.430  1.574e-03  1.593e-03  63.490  64.150
3 ... 0      1      -18.340 -17.980 -6.276e-04 -6.220e-04 -27.610 -26.970
4 ... 0      1      -29.070 -28.730 -4.628e-04 -4.519e-04 -17.430 -17.220
5 ... 0      1       -6.171  -5.867  1.343e-05  2.321e-05  -2.308  -2.115
6 ... 0      1       -8.770  -8.502 -1.026e-04 -1.043e-04 -13.520 -13.330
  y_RR      y_RRH    
1  94.97610  95.87767
2  79.52834  79.07920
3 -11.37910 -12.41159
4 -28.62189 -30.16971
5 -13.24151 -15.24268
6 -24.36453 -26.74102

## SAD models with matrix A

In [20]:
###############################################################################################
pheno4=read.table("dat_correct_RRA_RRH")



sln_sad <- read.table("sln_sadA", quote="\"") 
ide_sad<- read.table("ide_sadA", quote="\"")
sln_sad$ide_sad = ide_sad$V3


sln_sad$somme= sln_sad$V3 + sln_sad$ide_sad
names(sln_sad)<-c("V1","code","g","SEg","e","somme")



m_sadA <- as.data.frame(matrix( unlist( strsplit(as.character(sln_sad$code)  , "\\." ) ) , ncol = 2 , byrow = TRUE ))

sln_sad1=cbind(sln_sad,m_sadA)
names(sln_sad1)<-c("V1","code","g","SEgA","e","somme","temps","animal")
sln_sad1$temps=as.numeric(levels(sln_sad1$temps))[sln_sad1$temps]
sln_sad2=arrange(sln_sad1,animal,temps)
N_animal=length(unique(sln_sad2$animal))  #number animal from SAD.sln



sln_sad2$week=sln_sad2$temps
sln_sad2=subset(sln_sad2,select=-c(temps))
critere=c("animal","week")
DF_s=merge(pheno4,sln_sad2,by=c(critere))
DF_s$y_SAD = DF_s$e_SAD + DF_s$somme

DF_s <- select(DF_s, -c(somme,code, V1,g,e))

head(DF_s)
write.table(DF_s,"dat_correct_RRA_RRH_sadA")
############################################################


animal        week identif       bande  sex lignee gener loge eleve pds_dc
1 17MAG20005070 10   17MAG20005070 200011 2   n      G0    71   17MAG 25    
2 17MAG20005070  2   17MAG20005070 200011 2   n      G0    71   17MAG 25    
3 17MAG20005070  4   17MAG20005070 200011 2   n      G0    71   17MAG 25    
4 17MAG20005070  5   17MAG20005070 200011 2   n      G0    71   17MAG 25    
5 17MAG20005070  6   17MAG20005070 200011 2   n      G0    71   17MAG 25    
6 17MAG20005070  7   17MAG20005070 200011 2   n      G0    71   17MAG 25    
  ... e_RR    e_RRss   e_SAD      e_SADss    e_MTA   e_MTH   y_RR     
1 ...  -1.120  -0.9432 -2.249e-04 -2.081e-04  -5.578  -5.368 -36.77958
2 ...  74.510  74.8500  1.718e-03  1.719e-03  74.170  74.060  94.97610
3 ...  67.060  67.4300  1.574e-03  1.593e-03  63.490  64.150  79.52834
4 ... -18.340 -17.9800 -6.276e-04 -6.220e-04 -27.610 -26.970 -11.37910
5 ... -29.070 -28.7300 -4.628e-04 -4.519e-04 -17.430 -17.220 -28.62189
6 ...  -6.171  -5.8670  1.343e-05  2.321e-05  -2.308  -2.115 -13.24151
  y_RRH     SEgA   y_SAD    
1 -39.69193 11.780 -36.55022
2  95.87767  8.846  92.79172
3  79.07920  7.179  78.04357
4 -12.41159  7.112 -11.63063
5 -30.16971  7.418 -29.23046
6 -15.24268  8.000 -14.00499

In [21]:
###################################################
#rm(list = ls())


###############################################################################################
DF_s=read.table("dat_correct_RRA_RRH_sadA")

sln_sadH <- read.table("sln_sadH", quote="\"") #old is G_sad
ide_sadH <- read.table("ide_sadH", quote="\"")
sln_sadH$ide_sad = ide_sadH$V3

sln_sadH$somme= sln_sadH$V3 + sln_sadH$ide_sad
names(sln_sadH)<-c("V1","code","g","SEg","e","somme")

m_sadH <- as.data.frame(matrix(unlist( strsplit(as.character(sln_sadH$code), "\\." )) , ncol = 2 , byrow = TRUE ))


sln_sadH1=cbind(sln_sadH,m_sadH)
names(sln_sadH1)<-c("V11","code1","g1","SEgH","e1","somme1","temps","animal")

sln_sadH1$temps=as.numeric(levels(sln_sadH1$temps))[sln_sadH1$temps]
sln_sadH2=arrange(sln_sadH1,animal,temps)
N_animal=length(unique(sln_sadH2$animal))  #number animal from SAD.sln
sln_sadH2$week=sln_sadH2$temps
sln_sadH2=select(sln_sadH2,-c(temps))




critere=c("animal","week")
DF_s2 =merge(DF_s,sln_sadH2,by=c(critere))


DF_s2$y_SADH = DF_s2$e_SADss + DF_s2$somme1

# Delete some columns
DF_s2=select(DF_s2,-c(code1,somme1,V11,code1,g1,e1))
# write table
write.table(DF_s2,"dat_correct_RRA_RRH_sadA_SADH")

In [22]:
head(DF_s2)

animal        week identif       bande  sex lignee gener loge eleve pds_dc
1 17MAG20005070 10   17MAG20005070 200011 2   n      G0    71   17MAG 25    
2 17MAG20005070  2   17MAG20005070 200011 2   n      G0    71   17MAG 25    
3 17MAG20005070  4   17MAG20005070 200011 2   n      G0    71   17MAG 25    
4 17MAG20005070  5   17MAG20005070 200011 2   n      G0    71   17MAG 25    
5 17MAG20005070  6   17MAG20005070 200011 2   n      G0    71   17MAG 25    
6 17MAG20005070  7   17MAG20005070 200011 2   n      G0    71   17MAG 25    
  ... e_SAD      e_SADss    e_MTA   e_MTH   y_RR      y_RRH     SEgA  
1 ... -2.249e-04 -2.081e-04  -5.578  -5.368 -36.77958 -39.69193 11.780
2 ...  1.718e-03  1.719e-03  74.170  74.060  94.97610  95.87767  8.846
3 ...  1.574e-03  1.593e-03  63.490  64.150  79.52834  79.07920  7.179
4 ... -6.276e-04 -6.220e-04 -27.610 -26.970 -11.37910 -12.41159  7.112
5 ... -4.628e-04 -4.519e-04 -17.430 -17.220 -28.62189 -30.16971  7.418
6 ...  1.343e-05  2.321e-05  -2.308  -2.115 -13.24151 -15.24268  8.000
  y_SAD     SEgH   y_SADH   
1 -36.55022 11.920 -39.64021
2  92.79172  8.403  92.78072
3  78.04357  7.064  76.62459
4 -11.63063  7.190 -13.23762
5 -29.23046  7.678 -31.02945
6 -14.00499  8.401 -16.34498

## Multiple trait models

In [23]:
# rm(list = ls())

# Read files
DF_s2=read.table("dat_correct_RRA_RRH_sadA_SADH")
sln_mt <- read.table("sln_mtA", quote="\"", comment.char="") #old is G_sad
ide_mt <- read.table("ide_mtA", quote="\"", comment.char="")
sln_mt$V3=sln_mt$V3+ide_mt$V3
names(sln_mt)<-c("V1","code","value","SE")





m_mtA <- as.data.frame(matrix( unlist(strsplit(as.character(sln_mt$code)  , "\\.")) , ncol = 2 , byrow = TRUE ))
sln_mt1=cbind(sln_mt,m_mtA)
names(sln_mt1)<-c("V1","code","value","SEg_MTA","pe2","animal")
sln_mt1$pe2=as.numeric(levels(sln_mt1$pe2))[sln_mt1$pe2]
sln_mt2=arrange(sln_mt1,animal,pe2)
N_animal=length(unique(sln_mt2$animal))  #number animal from SAD.sln


critere=c("animal","pe2")
DF_s3=merge(DF_s2,sln_mt2,by=c(critere))
DF_s3$y_MT = DF_s3$e_MTA+DF_s3$value # corrected phenotypes
DF_s3 = select(DF_s3,-c(value, code, V1)) # drop some columns
write.table(DF_s3,"dat_correct_RRA_RRH_sadA_SADH_mtA")
########################################################################################################

head(DF_s3)


animal        pe2 week identif       bande  sex lignee gener loge eleve ...
1 17MAG20005070 1   2    17MAG20005070 200011 2   n      G0    71   17MAG ...
2 17MAG20005070 2   4    17MAG20005070 200011 2   n      G0    71   17MAG ...
3 17MAG20005070 2   5    17MAG20005070 200011 2   n      G0    71   17MAG ...
4 17MAG20005070 3   6    17MAG20005070 200011 2   n      G0    71   17MAG ...
5 17MAG20005070 3   7    17MAG20005070 200011 2   n      G0    71   17MAG ...
6 17MAG20005070 3   8    17MAG20005070 200011 2   n      G0    71   17MAG ...
  e_MTA   e_MTH   y_RR      y_RRH     SEgA  y_SAD     SEgH  y_SADH    SEg_MTA
1  74.170  74.060  94.97610  95.87767 8.846  92.79172 8.403  92.78072 7.612  
2  63.490  64.150  79.52834  79.07920 7.179  78.04357 7.064  76.62459 5.936  
3 -27.610 -26.970 -11.37910 -12.41159 7.112 -11.63063 7.190 -13.23762 5.936  
4 -17.430 -17.220 -28.62189 -30.16971 7.418 -29.23046 7.678 -31.02945 6.505  
5  -2.308  -2.115 -13.24151 -15.24268 8.000 -14.00499 8.401 -16.34498 6.505  
6 -13.520 -13.330 -24.36453 -26.74102 8.805 -25.16910 9.251 -27.83010 6.505  
  y_MT    
1  95.4500
2  80.2270
3 -10.8730
4 -28.2649
5 -13.1429
6 -24.3549

In [24]:
########################################################################################################
#rm(list = ls())



DF_s3=read.table("dat_correct_RRA_RRH_sadA_SADH_mtA")
sln_mtH <- read.table("sln_mtH", quote="\"", comment.char="") #old is G_sad
ide_mtH <- read.table("ide_mtH", quote="\"", comment.char="")
sln_mtH$V3=sln_mtH$V3+ide_mtH$V3
names(sln_mtH)<-c("V1","code","value","SE")




m_mtH <- as.data.frame(matrix(unlist(strsplit(as.character(sln_mtH$code)  , "\\." )), ncol = 2 , byrow = TRUE ))
sln_mtH1=cbind(sln_mtH,m_mtH)
names(sln_mtH1)<-c("V11","code1","value1","SEg_MTH","pe2","animal")
head(sln_mtH1)
sln_mtH1 <- select(sln_mtH1,c(animal,pe2,value1,SEg_MTH))

V11                  code1           value1  SEg_MTH pe2 animal       
1 period1.giv(animal,1 1.17M3019830268 0.10790 7.714   1   17M3019830268
2 period1.giv(animal,1 1.17M3019830280 0.02548 7.714   1   17M3019830280
3 period1.giv(animal,1 1.17M3019830492 0.04120 7.714   1   17M3019830492
4 period1.giv(animal,1 1.17M3019830508 0.04120 7.714   1   17M3019830508
5 period1.giv(animal,1 1.17M3019840010 0.09216 7.714   1   17M3019840010
6 period1.giv(animal,1 1.17M3019840036 0.08239 7.714   1   17M3019840036

In [25]:
critere=c("animal","pe2")
DF_s4=merge(DF_s3,sln_mtH1,by=c(critere))
DF_s4$y_MTH = DF_s4$e_MTH+DF_s4$value1

DF_s4=subset(DF_s4,select=-c(value1))
write.table(DF_s4,"dat_correct_RRA_RRH_sadA_SADH_mtA_mtH")


with(DF_s4,cor(y_RR,y_SADH))
with(DF_s4,cor(y_RRH,y_RR))
head(DF_s4)


[1] 0.9947062

[1] 0.9960819

animal        pe2 week identif       bande  sex lignee gener loge eleve ...
1 17MAG20005070 1   2    17MAG20005070 200011 2   n      G0    71   17MAG ...
2 17MAG20005070 2   4    17MAG20005070 200011 2   n      G0    71   17MAG ...
3 17MAG20005070 2   5    17MAG20005070 200011 2   n      G0    71   17MAG ...
4 17MAG20005070 3   6    17MAG20005070 200011 2   n      G0    71   17MAG ...
5 17MAG20005070 3   7    17MAG20005070 200011 2   n      G0    71   17MAG ...
6 17MAG20005070 3   8    17MAG20005070 200011 2   n      G0    71   17MAG ...
  y_RR      y_RRH     SEgA  y_SAD     SEgH  y_SADH    SEg_MTA y_MT     SEg_MTH
1  94.97610  95.87767 8.846  92.79172 8.403  92.78072 7.612    95.4500 6.915  
2  79.52834  79.07920 7.179  78.04357 7.064  76.62459 5.936    80.2270 5.822  
3 -11.37910 -12.41159 7.112 -11.63063 7.190 -13.23762 5.936   -10.8730 5.822  
4 -28.62189 -30.16971 7.418 -29.23046 7.678 -31.02945 6.505   -28.2649 6.555  
5 -13.24151 -15.24268 8.000 -14.00499 8.401 -16.34498 6.505   -13.1429 6.555  
6 -24.36453 -26.74102 8.805 -25.16910 9.251 -27.83010 6.505   -24.3549 6.555  
  y_MTH  
1  96.640
2  79.649
3 -11.471
4 -30.280
5 -15.175
6 -26.390

In [26]:
sessionInfo()


R version 3.6.3 (2020-02-29)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252 
[2] LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] corrplot_0.84   psych_1.9.12.31 plyr_1.8.5      stringr_1.4.0  
[5] reshape2_1.4.3  dplyr_0.8.4     pacman_0.5.1   

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.4       magrittr_1.5     mnormt_1.5-6     tidyselect_1.0.0
 [5] uuid_0.1-2       lattice_0.20-38  R6_2.4.1         rlang_0.4.5     
 [9] tools_3.6.3      grid_3.6.3       parallel_3.6.3   nlme_3.1-144    
[13] htmltools_0.4.0  digest_0.6.25    assertthat_0.2.1 tibble_2.1.3    
[17] crayon_1.3.4     IRdisplay_0.7.0  purrr_0.3.3    

In [27]:

#rm(list = ls())

pacman::p_load(dplyr,reshape2, stringr, plyr, psych, corrplot)


ped <- read.table("geneal.gen", quote="\"", comment.char="")
names(ped)= c("animal","pere","mere")


In [42]:
head(ped,20)

animal        pere          mere         
1  17M3019830268 0             0            
2  17M3019830280 0             0            
3  17M3019830492 0             0            
4  17M3019830508 0             0            
5  17M3019840010 17M3019830280 17M3019830268
6  17M3019840036 17M3019830508 17M3019830492
7  0000019851067 0             0            
8  0000019858909 0             0            
9  0000019858971 0             0            
10 0000019859155 0             0            
11 8600119854418 0             0            
12 0000019860769 0             0            
13 0000019860909 0             0            
14 23RD119864470 0             0            
15 29JEA19860829 0             0            
16 79KVS19860901 0             0            
17 17M3019860299 8600119854418 17M3019840010
18 17M3019860332 17M3019840036 17M3019830268
19 02JC119876571 0             0            
20 17BM119873406 0             0

In [36]:
ped1=arrange(ped,mere,animal)
ped2=arrange(ped,pere,animal)


In [43]:
head(ped1,20)

animal        pere mere
1  0000019851067 0    0   
2  0000019858909 0    0   
3  0000019858971 0    0   
4  0000019859155 0    0   
5  0000019860769 0    0   
6  0000019860909 0    0   
7  02JC119876571 0    0   
8  03HP119903200 0    0   
9  03HP119913663 0    0   
10 03HP119924391 0    0   
11 12ZAG19931898 0    0   
12 16HAC19920087 0    0   
13 16HAC19961200 0    0   
14 16HAC19963327 0    0   
15 16HAC19963563 0    0   
16 16HAC19965058 0    0   
17 16HAC19991454 0    0   
18 17BM119873406 0    0   
19 17M3019830268 0    0   
20 17M3019830280 0    0

In [44]:
head(ped2,20)

animal        pere mere
1  0000019851067 0    0   
2  0000019858909 0    0   
3  0000019858971 0    0   
4  0000019859155 0    0   
5  0000019860769 0    0   
6  0000019860909 0    0   
7  02JC119876571 0    0   
8  03HP119903200 0    0   
9  03HP119913663 0    0   
10 03HP119924391 0    0   
11 12ZAG19931898 0    0   
12 16HAC19920087 0    0   
13 16HAC19961200 0    0   
14 16HAC19963327 0    0   
15 16HAC19963563 0    0   
16 16HAC19965058 0    0   
17 16HAC19991454 0    0   
18 17BM119873406 0    0   
19 17M3019830268 0    0   
20 17M3019830280 0    0

In [46]:

#names(ped1)= c("progeny","animal","mari")
names(ped1)= c("progeny","mari","animal") # list classed by dam
names(ped2)= c("progeny","animal","wife") # ist classed by sire



In [47]:
head(ped1,20)

progeny       mari animal
1  0000019851067 0    0     
2  0000019858909 0    0     
3  0000019858971 0    0     
4  0000019859155 0    0     
5  0000019860769 0    0     
6  0000019860909 0    0     
7  02JC119876571 0    0     
8  03HP119903200 0    0     
9  03HP119913663 0    0     
10 03HP119924391 0    0     
11 12ZAG19931898 0    0     
12 16HAC19920087 0    0     
13 16HAC19961200 0    0     
14 16HAC19963327 0    0     
15 16HAC19963563 0    0     
16 16HAC19965058 0    0     
17 16HAC19991454 0    0     
18 17BM119873406 0    0     
19 17M3019830268 0    0     
20 17M3019830280 0    0

In [48]:
head(ped2,20)

progeny       animal wife
1  0000019851067 0      0   
2  0000019858909 0      0   
3  0000019858971 0      0   
4  0000019859155 0      0   
5  0000019860769 0      0   
6  0000019860909 0      0   
7  02JC119876571 0      0   
8  03HP119903200 0      0   
9  03HP119913663 0      0   
10 03HP119924391 0      0   
11 12ZAG19931898 0      0   
12 16HAC19920087 0      0   
13 16HAC19961200 0      0   
14 16HAC19963327 0      0   
15 16HAC19963563 0      0   
16 16HAC19965058 0      0   
17 16HAC19991454 0      0   
18 17BM119873406 0      0   
19 17M3019830268 0      0   
20 17M3019830280 0      0

In [64]:
list_genotype <- read.table("list", quote="\"")
list_genotype$g = 1
list_genotype$animal=list_genotype[,2]
list_g=list_genotype[,c("animal","g")]

head(list_genotype)

V1 V2            g animal       
1 0  17MAG20004838 1 17MAG20004838
2 1  17MAG20004841 1 17MAG20004841
3 2  17MAG20004842 1 17MAG20004842
4 3  17MAG20004965 1 17MAG20004965
5 4  17MAG20005004 1 17MAG20005004
6 5  17MAG20005005 1 17MAG20005005

In [65]:
list_g2= merge(ped1,list_g,c("animal")) # dams with genotypes
list_g2_p= merge(ped2,list_g,c("animal")) # sires with genotypes


In [66]:
head(list_g2,20)

animal        progeny       mari          g
1  17MAG20004838 17MAG20012561 17MAG20005775 1
2  17MAG20004838 17MAG20012567 17MAG20005775 1
3  17MAG20004838 17MAG20012560 17MAG20005775 1
4  17MAG20004838 17MAG20012562 17MAG20005775 1
5  17MAG20004841 8600120010023 17MAG20005197 1
6  17MAG20004841 8600120010021 17MAG20005197 1
7  17MAG20004841 8600120020008 17MAG20005197 1
8  17MAG20004841 8600120020010 17MAG20005197 1
9  17MAG20004841 8600120010026 17MAG20005197 1
10 17MAG20004842 8600120012045 17MAG20005350 1
11 17MAG20004842 8600120012043 17MAG20005350 1
12 17MAG20004842 8600120012041 17MAG20005350 1
13 17MAG20004842 8600120012042 17MAG20005350 1
14 17MAG20004965 17MAG20024167 17MAG20005197 1
15 17MAG20004965 17MAG20012571 17MAG20005197 1
16 17MAG20004965 17MAG20012572 17MAG20005197 1
17 17MAG20004965 17MAG20012574 17MAG20005197 1
18 17MAG20005004 8600120020032 17MAG20005629 1
19 17MAG20005004 8600120020029 17MAG20005629 1
20 17MAG20005004 8600120020028 17MAG20005629 1

Herein, we observed that the dam 17MAG20004838 has 4 progenies : <br>
- 17MAG20012561
- 17MAG20012567
- 17MAG20012560
- 17MAG20012562

In [67]:
head(list_g2_p,28)
list_g2_p[1:23,2]

animal        progeny       wife          g
1  17MAG20005073 8600120022019 17MAG20005322 1
2  17MAG20005073 8600120022110 17MAG20005316 1
3  17MAG20005073 8600120022018 17MAG20005322 1
4  17MAG20005073 8600120022015 17MAG20005322 1
5  17MAG20005073 17MAG20012765 17MAG20005554 1
6  17MAG20005073 17MAG20012768 17MAG20005554 1
7  17MAG20005073 8600120022016 17MAG20005322 1
8  17MAG20005073 8600120022111 17MAG20005316 1
9  17MAG20005073 17MAG20013046 17MAG20005601 1
10 17MAG20005073 8600120012005 17MAG20005322 1
11 17MAG20005073 17MAG20012762 17MAG20005554 1
12 17MAG20005073 8600120012163 17MAG20005478 1
13 17MAG20005073 8600120012166 17MAG20005478 1
14 17MAG20005073 17MAG20012763 17MAG20005554 1
15 17MAG20005073 8600120012160 17MAG20005478 1
16 17MAG20005073 17MAG20013047 17MAG20005601 1
17 17MAG20005073 17MAG20013045 17MAG20005601 1
18 17MAG20005073 8600120012141 17MAG20005440 1
19 17MAG20005073 8600120012142 17MAG20005440 1
20 17MAG20005073 8600120012003 17MAG20005322 1
21 17MAG20005073 8600120012010 17MAG20005322 1
22 17MAG20005073 8600120012157 17MAG20005478 1
23 17MAG20005073 8600120012138 17MAG20005440 1
24 17MAG20005197 17MAG20012751 17MAG20005533 1
25 17MAG20005197 17MAG20012571 17MAG20004965 1
26 17MAG20005197 17MAG20012750 17MAG20005533 1
27 17MAG20005197 8600120020045 17MAG20005441 1
28 17MAG20005197 17MAG20012483 17MAG20005154 1

[1] 8600120022019 8600120022110 8600120022018 8600120022015 17MAG20012765
 [6] 17MAG20012768 8600120022016 8600120022111 17MAG20013046 8600120012005
[11] 17MAG20012762 8600120012163 8600120012166 17MAG20012763 8600120012160
[16] 17MAG20013047 17MAG20013045 8600120012141 8600120012142 8600120012003
[21] 8600120012010 8600120012157 8600120012138
3986 Levels: 0000019851067 0000019858909 0000019858971 ... FI04319920706

Herein, we observed that the sire 17MAG20004838 has 23 progenies : <br>
- 8600120022019
- 8600120022110
- 8600120022018
- 8600120022015
- 17MAG20012765
- ...
- 8600120012138

In [68]:
length(unique((list_g2$animal)))
length(unique((list_g2_p$animal)))
names(list_g2)= c("mere","animal","pere","g")
names(list_g2_p)= c("pere","animal","mere","g")

[1] 568

[1] 80

- 568 dams with genotypes
- 80 sires with genotypes

In [ ]:
DF_s4 <- read.table("dat_correct_RRA_RRH_sadA_SADH_mtA_mtH", header=TRUE, quote="\"")
df=merge(list_g2,DF_s4,c("animal"))



length(df$animal)
df1=arrange(df,mere,animal,week)
df2=subset(df1,!is.na(week))

df_p=merge(list_g2_p,DF_s4,c("animal"))
length(df_p$animal)
df1_p=arrange(df_p,pere,animal,week)
df2_p=subset(df1_p,!is.na(week))


df2$animal=as.character(df2$animal)
df2$mere=as.character(df2$mere)
df2$week=as.numeric(df2$week)



In [ ]:


df3=select(df2,animal,mere,week,y_RR,y_RRH,y_SAD,y_SADH,y_MT,y_MTH, FI, FCR, ADG, RFI)

df3$sub <- paste0(df3$week,df3$mere)
df3 <- arrange(df3,sub)


w=aggregate(y_RR ~ sub, data = df3, FUN=function(x) length(x))[,2]

df3$w=rep(w,w)


df4=arrange(df3,mere,week)
df4 <- df4[,-1]
# df4=aggregate(. ~ week, data = df3, FUN=function(x) mean(x))
df4$animal=df4$mere

df5=arrange(df4,animal,week)





df2_p$animal=as.character(df2_p$animal)

df3_p=select(df2_p,animal,pere,week,y_RR,y_RRH,y_SAD,y_SADH,y_MT,y_MTH,FI, FCR, ADG, RFI)
df3_p$sub <- paste0(df3_p$week,df3_p$pere)
df3_p <- arrange(df3_p,sub)



N=aggregate(y_RR ~ sub, data = df3_p, FUN=function(x) length(x))[,2]
df3_p$w=rep(N,N)

df4_p <- arrange(df3_p,pere,week)
df4_p <- df4_p[,-1]

df4_p$animal=df4_p$pere
df4_p=arrange(df4_p,animal,week)

df6_p=select(df4_p,animal,animal,week,y_RR,y_RRH,y_SAD,y_SADH,y_MT,y_MTH,FI, FCR, ADG, RFI,w)

df6=select(df5,animal,animal,week,y_RR,y_RRH,y_SAD,y_SADH,y_MT,y_MTH,FI, FCR, ADG, RFI,w)

df6_p$l=2
df6$l=1

df7=rbind(df6,df6_p)

df8=arrange(df7,animal,week)










# names(df5)<-c("animal","week","y_RR","y_RRH")
# df5$l=1

cory$week=cory$temps.y
cory1=select(cory,animal,week,y_RR,y_RRH,y_SAD,y_SADH,y_MT,y_MTH,FI, FCR, ADG, RFI)
cory1=arrange(cory1,animal,week)
cory1$l=3
cory1$w=1

df9=rbind(df8,cory1)


df9=merge(cory1,df8,c("animal","week","y_RR","y_RRH","y_SAD","y_SADH","y_MT","y_MTH","FI", "FCR", "ADG", "RFI","w","l"),all=T)
df10=arrange(df9,animal,week)

ped_all <- read.csv("pedigreeRFIall", header=F, sep="")

names(ped_all)<-c("N","lignee","gener", "animal","p","m")

df_all=merge(ped_all,df10,c("animal"))


df_all1=arrange(df_all,animal,week)

df_all2=select(df_all1,animal,week,gener,lignee,y_RR,y_RRH,y_SAD,y_SADH,y_MT,y_MTH,FI, FCR, ADG, RFI,w,l)

df_all2=arrange(df_all2,animal,week,w)


df_all2$animal=as.character(df_all2$animal)
df_all2$week=as.numeric(df_all2$week)
df_all2$lignee=as.character(df_all2$lignee)
df_all2$gener=as.character(df_all2$gener)


df_all3=aggregate(y_RR ~animal+week+gener+lignee+w+l, data = df_all2, mean)
df_all3$y_RRH=aggregate(y_RRH ~animal+week+gener+lignee+w+l, data = df_all2, mean)[,7]
df_all3$y_SAD=aggregate(y_SAD ~animal+week+gener+lignee+w+l, data = df_all2, mean)[,7]
df_all3$y_SADH=aggregate(y_SADH ~animal+week+gener+lignee+w+l, data = df_all2, mean)[,7]
df_all3$y_MT=aggregate(y_MT ~animal+week+gener+lignee+w+l, data = df_all2, mean)[,7]
df_all3$y_MTH=aggregate(y_MTH ~animal+week+gener+lignee+w+l, data = df_all2, mean)[,7]
df_all3$y_FI=aggregate(FI ~animal+week+gener+lignee+w+l, data = df_all2, mean)[,7]
df_all3$y_FCR=aggregate(FCR ~animal+week+gener+lignee+w+l, data = df_all2, mean)[,7]
df_all3$y_RFI=aggregate(RFI ~animal+week+gener+lignee+w+l, data = df_all2, mean)[,7]
df_all3$y_ADG=aggregate(ADG ~animal+week+gener+lignee+w+l, data = df_all2, mean)[,7]




df_all4=arrange(df_all3,animal,week)

write.table(df_all4,"DATA_corrected.txt")
